app/src/main/java/com/github/polypoly/app/base/game/Player.kt

```diff
@@ -20,9 +17,8 @@ import kotlin.random.Random
 data class Player (
     val user: User = User(),
     private var balance: Int = 0,
-    private var ownedLocations: List<InGameLocation> = listOf(),
+    private var ownedLocations: MutableList<InGameLocation> = mutableListOf(),
```

Comment:
For the tests, I have to either do that or put them private. If you have another solution, I'm fine, but I don't think putting it mutable doesn't make sense given that the player's own locations change during the game.

Reply:
I see your point, due to the lack of a proper way to inject the player as a dependency of `GameViewModel` it is hard to do the testing...
An somewhat acceptable and easy tradeoff could be the following:
- Leave the attribute of Player as a `MutableList` as you did: it is in fact not such a big issue as the player is initialized by the `Game` object when using `launchFromPendingGame` so the actual usage remains safe
- Open the constructor of Game to allow direct instantiation (it is private at the moment): in the `init` of `GameActivityLandlordTest` you can then initialize an attribute storing a reference of the player's list of locations owned that you can then use later in your tests
- Make `getOwnedLocations` in `Player` return an immutable copy of the list of locations: this is to me the biggest encapsulation issue here as it returns the `MutableList` of locations stored directly
This would give something like that:
```kotlin
    val playerProperties: MutableList<InGameLocation> = mutableListOf()

    init {
        GlobalInstances.currentUser = TEST_USER_0
        GameRepository.player = Player(TEST_USER_0, 100, playerProperties) // May not be needed after merge with future pr
        GameRepository.game = Game(
            TEST_USER_0,
            listOf(
                GameRepository.player!!,
                Player(TEST_USER_1, 100, mutableListOf())
            ),
            GameParameters(GameMode.LANDLORD, 2, maxRound = 10)
        )
    }

    ...

    @Before
    fun setUp() {
        playerProperties.clear()
    }

    ...

    @Test
    fun whenClickingOnTradeYouCanSeeAPopUpWithBuildingsToChoose() {
        // give some location to the player
        for(i in 0..2) {
            val inGameLocation = GameRepository.game?.getInGameLocation()?.get(i)!! // Should be able to use PolyPolyTest.getRandomLocations() instead
            playerProperties.add(inGameLocation) // we use the attribute that is shared with the player
        }

        composeTestRule.onNodeWithTag("other_players_and_game_hud").performClick()
        composeTestRule.onAllNodesWithTag("other_player_hud")[0].performClick()
        composeTestRule.onNodeWithTag("trade_button").performClick()
        composeTestRule.onNodeWithTag("locations_list_dialog").assertIsDisplayed()
        composeTestRule.onNodeWithText("Choose a location to trade").assertIsDisplayed()
        composeTestRule.onNodeWithText("Cancel").assertIsDisplayed()
    }
```
This is still far from being perfect, but again without a dependency injection framework I indeed don't find a perfectly clean way to solve this problem... Here at least the encapsulation issue is left only at initialization while the rest of the interface is more concise and robust so it could be considered acceptable as in our real code the initialization is robust trough the use of `Game.launchFromPendingGame`
This may look like a lot to change for not much, but in fact it is only a matter of a couple of lines to change, and `GameRepository.player` should disappear in a future refactoring pr, so some parts here will have to be changed anyway...

Reply:
thank you, I have done something similar

Reply:
I see you put the tradeoff by exposing the `player` in `getAdmin`, that works even better!